### PR TITLE
1.0 tls support

### PR DESCRIPTION
**Pull request by jranson**
This builds on #296, which provides basic TLS support into `master` thanks to @u5surf.  This PR extends that functionality into `next`, and provides additional functionality. In this port+extension, multiple TLS configs can be provided, and each origin configuration can be mapped to any one of the configured TLS sections. Further, each origin can support http only, http+https, or https only, depending upon the operator's preference. Trickster will only listen on the TLS and/or HTTP ports if origins are configured for either, and origins will only be routed to the TLS and/or HTTP router based on the operating configuration. Test coverage is also included, as well as a few ancillary bug fixes that were identified along the way during testing.

### CODE REVIEW

**Comment by jranson** [comment]
this PS is about 92.7% completed at this point, but I forgot to mark the 'Draft' flag on it. i will request reviews when it is fully ready.

**Comment by jranson** [comment]
PR ready for review.

File: .gitignore [hunk 1]
```diff
@@ -31,9 +31,6 @@ OPATH
 ehthumbs.db
 Thumbs.db
 
-# The binary
-trickster
```
Comment (LimitlessEarth) [inline .gitignore]: Do we want to ignore it in OPATH as well?

File: cmd/trickster/main.go [hunk 2]
```diff
@@ -72,12 +73,47 @@ func main() {
...
+					config.ProxyServer.ConnectionsLimit,
+					tlsConfig)
+				if err == nil {
+					err = http.Serve(l, handlers.CompressHandler(routing.TLSRouter))
+				}
+			}
+			log.Error("exiting", log.Pairs{"err": err})
+			wg.Done()
+		}()
+	}
```
Comment (LimitlessEarth) [inline cmd/trickster/main.go]: Do we want to log here in the else condition? Want to avoid failing silently.

File: .gitignore (hunk 1 above)
Comment (jranson) [reply .gitignore]: OPATH is already in .gitignore so we're good

File: cmd/trickster/main.go (hunk 2 above)
Comment (jranson) [reply cmd/trickster/main.go]: I think we are OK here, in that it should not be possible to fail silently. The condition block starting L66 will fatal out before getting to this point if no listeners are configured, and the configuration loader L48 will fail out earlier than that if TLS is misconfigured; therefore, at this point of the execution, at least one of the listener blocks must execute and they are both logging errors.

**Pull request merged**